aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -67,77 +137,64 @@ namespace detail {
-    #endif // __HIP_PLATFORM_HCC__
-    return internals;
+  // Helper to determine the index of the stream to return
+  // Streams are returned round-robin, and the counter is "kept" between
+  // 0 and STREAMS_PER_POOl. 
+  // Note: it is possible the counter will grow beyond STREAMS_PER_POOL 
+  // temporarily, but it will eventually return to the desired range.
+  static int get_idx(std::atomic<int> &counter) {
+    int raw_idx = counter++;
+    int modded = raw_idx % STREAMS_PER_POOL;
+    if (raw_idx >= STREAMS_PER_POOL && modded == 0) {
+      counter -= STREAMS_PER_POOL;
+    }
+    return modded;
   }
 
-  // Note: despite not being "unsafe," is using these methods in a multithreaded
-  // environment then the caller must be sure that streams are valid
-  // when they're requested. These methods will throw an error if an
-  // invalid stream is requested.
-  CUDAStreamInternals* CUDAStream_getAndRetainCurrentStreamOnDevice(int64_t device) {
+  CUDAStreamInternals* CUDAStream_getDefaultStream(int64_t device) {
     initCUDAStreamsOnce();
+    if (device == -1) device = current_device();
     check_gpu(device);
-    auto cur = current_streams[device];
-    AT_ASSERT(CUDAStream_retain(cur));
-    return cur;
-  }
-  CUDAStreamInternals* CUDAStream_getAndRetainCurrentStream() {
-    return CUDAStream_getAndRetainCurrentStreamOnDevice(current_device());
+    return &default_streams[device];
   }
 
-  // Note: these unsafe methods do not retain the stream before returning it.
-  // This is unsafe behavior and these methods SHOULD NOT BE USED.
-  // They are here only for legacy compatibility.
-  CUDAStreamInternals* CUDAStream_getCurrentStreamOnDeviceUnsafe(int64_t device) {
+  // Returns a stream from the requested pool
+  // Note: when called the first time on a device, this will create the
+  // stream pools for that device.
+  CUDAStreamInternals* CUDAStream_createStream(
+    const bool isHighPriority
+  , int64_t device) {
     initCUDAStreamsOnce();
+    if (device == -1) device = current_device();
     check_gpu(device);
-    return current_streams[device];
-  }
-  CUDAStreamInternals* CUDAStream_getCurrentStreamUnsafe() {
-    return CUDAStream_getCurrentStreamOnDeviceUnsafe(current_device());
-  }
+    
+    // Initializes the stream pools (once)
+    std::call_once(device_flags[device], initDeviceStreamState, device);
 
-  void CUDAStream_setStreamOnDevice(int64_t device, CUDAStreamInternals* ptr) {
-    initCUDAStreamsOnce();
-    check_gpu(device);
-    AT_ASSERT(ptr);
-    AT_ASSERT(ptr->device == device);
-    AT_ASSERT(CUDAStream_retain(ptr));
+    if (isHighPriority) {
+      const auto idx = get_idx(high_priority_counters[device]);
+      return &high_priority_streams[device][idx];
+    }
 
-    CUDAStream_free(current_streams[device]);
-    current_streams[device] = ptr;
-  }
+    const auto idx = get_idx(low_priority_counters[device]);
+    return &low_priority_streams[device][idx];
```

Comment:
Could we simply assign `low_priority_streams` or `high_priority_streams` to a reference? There's really no reason to duplicate the code.